utils/src/rational.rs

```diff
@@ -71,6 +94,62 @@ impl BigRationalExt for BigRational {
     fn from_frac_usize(numerator: usize, denominator: usize) -> Self {
         BigRational::new(BigInt::from(numerator), BigInt::from(denominator))
     }
+
+    fn log2_ceil(&self, binary_digits: usize) -> BigRational {
+        if self <= &BigRational::zero() {
+            panic!("log2 undefined for non-positive numbers");
+        }
```

Comment:
Should we also check that `binary_digits` > 0? (or accept NonZeroUsize)

Reply:
I think the behavior should be correct for binary_digits = 0, returning an integer result